File: 2331/EvaluateBooleanBinaryTree.cpp
/*
https://leetcode.com/problems/evaluate-boolean-binary-tree/description/?envType=daily-question&envId=2024-05-16
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    bool dfs(TreeNode *node){
        if(node->val==0) return false;
        if(node->val==1) return true;
        bool lft=false, rght=false;
        if(node!=nullptr){
            lft=dfs(node->left);
            rght=dfs(node->right);
            if(node->val==2){
                return lft||rght;
            } else{
                return lft&&rght;
            }
        }
        return false;
    }

    bool customPostOrder(TreeNode *node){
        return dfs(node);
    }

    bool evaluateTree(TreeNode* root) {
        if(root->val==0) return false;
        else if(root->val==1) return true;
        else{
            bool lft=customPostOrder(root->left);
            bool rght=customPostOrder(root->right);
            if(root->val==2) return lft||rght;
            else return lft&&rght;
        }
        return false;
    }
};

File: 2331/EvaluateBooleanBinaryTree.java
/*
https://leetcode.com/problems/evaluate-boolean-binary-tree/description/?envType=daily-question&envId=2024-05-16
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public boolean depthFirstSearchTraversal_forTrees(TreeNode node){
        if(node.val==0) return false;
        if(node.val==1) return true;
        boolean leftTreeTraversalValueForCurrentNode=false, rightTreeTraversalValueForCurrentNode=false;
        if(node!=null){
            leftTreeTraversalValueForCurrentNode=depthFirstSearchTraversal_forTrees(node.left);
            rightTreeTraversalValueForCurrentNode=depthFirstSearchTraversal_forTrees(node.right);
            if(node.val==2){
                return leftTreeTraversalValueForCurrentNode||rightTreeTraversalValueForCurrentNode;
            } else{
                return leftTreeTraversalValueForCurrentNode&&rightTreeTraversalValueForCurrentNode;
            }
        }
        return false;
    }

    public boolean customPostOrderTraversal(TreeNode node){
        return depthFirstSearchTraversal_forTrees(node);
    }

    public boolean evaluateTree(TreeNode root) {
        if(root.val==0) return false;
        else if(root.val==1) return true;
        else{
            boolean leftTreeTraversalForRoot=customPostOrderTraversal(root.left);
            boolean rightTreeTraversalForRoot=customPostOrderTraversal(root.right);
            if(root.val==2) return leftTreeTraversalForRoot||rightTreeTraversalForRoot;
            return leftTreeTraversalForRoot&&rightTreeTraversalForRoot;
        }
    }
}